### DISCUSSION TITLE
limbo balance not recovered after local force close

### DISCUSSION THREAD

**Discussion by nixonrichard1971** [lightningnetwork/lnd#7871] [Troubleshooting]
LND version: v0.16.2-beta

Issue: After a local force close of a channel (initiated by my node, possibly due to HTLCs?) which tx was confirmed >2weeks, one of the outputs is locked as "limbo balance" and my node is now creating lots of sweeps making many of my utxo's show as unconfirmed.

When i run the command `lncli pendingchannels` I get the pending close channel as output and the weird thing is that in anchor the result shows "LOST".

Any ideas on how to solve this? I have seen other discussions but I don't see this one matching any other.

**Comment by nixonrichard1971** [comment]
hi, any thoughts on this item, please? thank you

**Reply by guggero** [reply]
What chain backend? Neutrino?

**Reply by nixonrichard1971** [reply]
yes, in my lnd.conf file it says bitcoin.node=neutrino

**Reply by guggero** [reply]
LOST on an anchor just means that someone else pulled it in the meantime (which is okay, that's what anchors are there for, so that _someone_ can attach fees and confirm the transaction). Can you post the output of `lncli pendingchannels` please? Without that there's not much we can say.

**Reply by nixonrichard1971** [reply]
sure. here's the output:

```
{
    "total_limbo_balance": "9996530",
    "pending_open_channels": [
    ],
    "pending_closing_channels": [
    ],
    "pending_force_closing_channels": [
        {
            "channel": {
                "remote_node_pub": "0286b2daec36414f1e67fed886a58c4e650b273a9ed99b18d069a79eca4ff1a954",
                "channel_point": "af2d615d4422223348cc5e87049f43cafed76f4a22649b09854616bd7ec70cf8:0",
                "capacity": "10000000",
                "local_balance": "9996530",
                "remote_balance": "0",
                "local_chan_reserve_sat": "0",
                "remote_chan_reserve_sat": "0",
                "initiator": "INITIATOR_LOCAL",
                "commitment_type": "ANCHORS",
                "num_forwarding_packages": "0",
                "chan_status_flags": "",
                "private": false
            },
            "closing_txid": "b2934b6c681007caa7fb5d9f53d37f94883938bc9b3e5c04a79912992a2a9363",
            "limbo_balance": "9996530",
            "maturity_height": 0,
            "blocks_til_maturity": 0,
            "recovered_balance": "0",
            "pending_htlcs": [
            ],
            "anchor": "LOST"
        }
    ],
    "waiting_close_channels": [
    ]
}
```

**Reply by guggero** [reply]
Thanks. Yeah, it looks like `lnd` got confused by the anchor being pulled by another transaction. I think this will be fixed in the next major version. In your case, I think starting `lnd` with `reset-wallet-transactions=true` once and then let it re-sync fully (this can take many hours on Neutrino, be sure to not abort it, as it won't properly continue after a restart, check the log for progress).

**Reply by nixonrichard1971** [reply]
thanks. I'll try this. I've heard many people complaining about Neutrino backend. Would switching to Bitcoind make any difference?

**Reply by guggero** [reply]
Some cases are difficult for Neutrino, because when publishing a transaction, a Neutrino node doesn't get any feedback if the transaction is okay. So in some cases (especially around relay fee) that makes it tricky. Thought I think some of those cases will also be addressed with `lnd v0.17.0` which will come out soon.
But if you have the disk space it's definitely better to run `bitcoind`, even if it's just a pruned node.

**Reply by nixonrichard1971** [reply]
thanks. I have tried to restart lnd with reset-wallet-transactions=true in the lnd.conf file but it seems to be stuck. I have checked the lnd.log file and I just get lines just with this type of message at the beginning: 

`"2023-08-25 16:43:24.899 [ERR] BTCN: filter for block 000000000000000000000b6f60ea4c3a44e310b9f0b5eb315bbce7f246f35c23 might be invalid, input 0 of tx 4eb8c25de0c9520b9da9fd33140d631844f2fd340f7d68c8d5635786964c9ccd spends pk script 002057def57831dd7dd4f143b79b405b9f8e707f3cdbff6918a9cf01cbb4c1d9ccaf which wasn't matched by filter. The input likely spends a taproot output which is not yetsupported" `

and also towards the end of the log this type:

`"2023-08-25 16:41:07.279 [ERR] RPCS: [/chainrpc.ChainNotifier/RegisterBlockEpochNtfn]: chain notifier RPC is still in the process of starting"`

do you know what's wrong and/or how to solve this? I am afraid to reset it as you mentioned this could create problems.

**Reply by guggero** [reply]
As mentioned, this can take a couple of days, depending on your internet connection and number of peers. Both errors are nothing to worry about. What does `lncli getinfo` say?

**Reply by nixonrichard1971** [reply]
thanks. it just doesn't work as it did last time i made a reset-wallet-transactions restart. it was pretty neat and relatively fast (a matter of half a day and I could see how my wallet was developing having tx's being processed and updating utxos owned and subsequently updating balance). Now, this time, it seems it is in the same place it was when restarted (synchronizing) any command I use with lncli just hangs and doesn't give me any output so I have to close the terminal and open it again if i want to re-connect to my node.

**Reply by guggero** [reply]
Maybe it got stuck somewhere because it couldn't connect to any Neutrino peers. You can try switching to bitcoind or just trying again (with `reset-wallet-transactions=true` still being set, it will just start from the beginning again).

**Reply by nixonrichard1971** [reply]
I have restarted and it seemed to be loading correctly, but is stuck now again showing this error in the log:

2023-08-30 05:15:43.343 [ERR] RPCS: unable to populate arbitrator force close resp:af2d615d4422223348cc5e87049f43cafed76f4a22649b09854616bd7ec70cf8:0, unable to find arbitrator

this has never happened before. I can see it's a tx that was confirmed 4 months ago.

**Reply by guggero** [reply]
Does the channel still show in `pending_force_closing_channels`? It can be that it is still re-scanning the chain to attempt to resolve it.
An alternative to waiting is the [`chantools sweeptimelockmanual`](https://github.com/lightninglabs/chantools/blob/master/doc/chantools_sweeptimelockmanual.md) command. The info required to call that can be extracted from the channel backup (be sure to read the help text of the command linked).

**Reply by nixonrichard1971** [reply]
when i call `lncli pendingchannels` it shows me `[lncli] rpc error: code = Unknown desc = unable to find arbitrator`. in this case if I run lncli get info it throws me an output, unlike last time (before I rebooted my node) that was stuck.

I have located the channel.backup file, but how do I read this to make sure to input the necessary data to the command?

Also just to confirm, I have looked into it and this seems to be the channel that was showing pending close due to the anchor being LOST that caused the problem in the first case.

Can you confirm I should try to run this command and if you can provide me just a bit of context so I run it successfully? I have checked the help but I'm not sure I fully understand what I need to do.

**Reply by guggero** [reply]
> I have located the channel.backup file, but how do I read this to make sure to input the necessary data to the command?

Did you read the link I sent you?

> Sweep the locally force closed state of a single channel manually if only a channel backup file is available. This can only be used if a channel is force closed from the local node but then that node's state is lost and only the channel.backup file is available.

> To get the value for --remoterevbasepoint you must use the dumpbackup command, then look up the value for RemoteChanCfg -> RevocationBasePoint -> PubKey.

**Reply by nixonrichard1971** [reply]
I understand I have to run a command like this (please let me know if incorrect):
`chantools sweeptimelockmanual
	--sweepaddr <one receiving address from my node's wallet>
	--timelockaddr < address of the time locked commitment output where the funds are stuck in, in this case is not the
	output of the openchannel tx, but the final address to which the input was paid to when the channel was force closed> 
	--remoterevbasepoint <the pubkey? or what should go here? and how do I recognize which one will it be since I will
	have many channels in the backup file?> 
	--feerate 10
	--publish`

so yes, I tried to read this and understand it, but is rather complex. First of all, how do I read the info I need to input from the channel.backup file using the terminal? I tried `dumpbackup <filepath of channel.backup>` and `lncli dumpbackup <filepath of channel.backup>` and none of them work so I find the --remoterevbasepoint input for the command.

**Reply by guggero** [reply]
The command you need to run is `chantools dumpbackup --multi_file <filepath of channel.backup>`. It will then ask you for the seed, because the backup is encrypted.
You can find the channel by searching for `FundingOutpoint` which should be `af2d615d4422223348cc5e87049f43cafed76f4a22649b09854616bd7ec70cf8:0`.

**Reply by nixonrichard1971** [reply]
Hi, thanks for that! I have now made progress and installed chantools successfully and found the input from the channel.backup for `--remoterevbasepoint`. Please note that the Pubkey I found starts with 02 and not with 03 as in the example, is that ok?

Also, can you please let me know what should I input for `--sweepaddr` and for `--timelockaddr` ? I have read the help but I am not completely sure if for the first one is any receiving address from my node's wallet, and for the second one is it the address of the open channel tx, or is the final address to which the input from the force closed was paid to? If you could confirm these two I'll try that and see if that solves this problem.

So, in this case, `--timelockaddr` I believe if I am not wrong looking into the closing tx b2934b6c681007caa7fb5d9f53d37f94883938bc9b3e5c04a79912992a2a9363 should be the output address with 9996530sats which is bc1qds72jpdkqxmfnuzzgm7e3w90wyjpusmh4nxr5uaefy7uzgmnz8us66c4cw.

Can you confirm, please?

**Reply by guggero** [reply]
Yes, public keys either start with 02 or 03, so that's fine. `--sweepaddr` is a new address from your wallet where the funds will be sent to if successful. So make sure to generate a new one that you control.
Correct, `--timelockaddr=bc1qds72jpdkqxmfnuzzgm7e3w90wyjpusmh4nxr5uaefy7uzgmnz8us66c4cw`.
Don't forget to also set a `--feerate` that works for the current mempool. Looks like at least 15 is recommended currently.

You can try running it with these parameters and it will try to find the correct keys locally, which can take a while. And depending on the number of channels your node has/had and the number of channel updates, it can be that you need to tweak some of the other parameters. But I'd just try with the defaults first.

**Reply by nixonrichard1971** [reply]
hi, thanks. I've tried two times with these inputs but it didn't seem to find the correct keys locally. First time it went over 240keys and the pc went to sleep and I am not sure if it went over all of them because my pc connection to the node broke with the pc going to sleep. The second time it went over the 500 key iteration and didn't find the correct keys (target script not derived). do I need to tweak any of the other parameters?

**Reply by guggero** [reply]
Yes, you probably need to tweak some parameters. Can you tell me what the `RemoteChanCfg -> RevocationBasePoint -> Path` value is for the channel (when running `chantools dumpbackup`)? And how old was the channel approximately? And was it very active (a lot of forwards or re-balances)?

**Reply by nixonrichard1971** [reply]
The path is "m/1017'/0'/0'/0/0" and the channel funding tx was confirmed 4months ago. The channel was not very active.

**Reply by guggero** [reply]
> The path is "m/1017'/0'/0'/0/0"

Sorry, my mistake. The remote path is always all zeroes at the end... What's the path for the local one is what we need to know in this case.

**Reply by nixonrichard1971** [reply]
the `LocalChanCfg -> RevocationBasePoint -> Path` is "m/1017'/0'/1'/0/170"

**Comment by guggero** [comment]
Okay, can you try with the following additional parameters (keep everything else as before): `--maxnumchanstotal=171 --maxnumchanupdates=5000`.

If that doesn't work, then it's possible the _other_ side closed the channel (hard to say from just the transaction itself). In that case it would be worth trying out `chantools sweepremoteclosed --recoverywindow 1000 --feerate 15 --sweepaddr <address_from_your_wallet_to_send_funds_to>`.

By the way, both these commands can be run on your computer as well, which might be faster. All you need is the correct binary for your operating system and the seed.

**Reply by nixonrichard1971** [reply]
thanks. For the first one with the additional parameters, it stopped at 171 keys and then threw and error:

`2023-09-04 22:04:00.744 [INF] CHAN: Tried 171 of 500 keys.
Error: target script not derived`

Prior to the reset-wallet-transactions I could see in Thunderhub that the channel was force closed locally (however i did not force-close it). do you think is possible that LND got confused with who closed this channel? I'll try and let you know if this solved the problem.

**Reply by nixonrichard1971** [reply]
Hi, just getting back with this one. The second script `chantools sweepremoteclosed` was run and it found one unspent output for one address (not the one we're looking for) and threw me this output:

```
2023-09-05 06:33:52.233 [INF] CHAN: Found 1 unspent outputs for address bc1qx3eyqqeucqwv84h65rzgnxtg8u0tmew4zpa9f0
2023-09-05 06:40:00.010 [INF] CHAN: Fee 1646 sats of 2680444 total amount (estimated weight 439)
2023-09-05 06:40:00.012 [INF] CHAN: Transaction: 02000000000101226d5cd84a56f5314323f35f259680b835897b45c23a209434f5c46f519de1e10100000000ffffffff010ee0280000000000160014608af4b83b932d0274e6421aeee3e4d337a0be0e024730440220169361474d236e955d21a20e226659547063ece74d88f9408e560e1ba06a977602206fb8a988545956e7a1e39975be1a4568bcfdbe63ffc9523245ceb585b386db86012102da2713e494de094a2cafac68a4d5ac2ac1ab51fd15673c47b6d660bf3652cf6d00000000
```
If I look for the address bc1qx3eyqqeucqwv84h65rzgnxtg8u0tmew4zpa9f0 (not the one we were concerned here anyway) the output has not been spent, so I am not sure if the tx has been published? FYI, now that I think of it I did not add `--publish` in this command (because I just copied your command above) so that might be why it did not publish it? But anyway, this seems to be a 3M sats channel that my peer has closed remotely possibly because I've been offline for too long already, and not the channel of 10M sats I'm having the issue with (lost anchor).

So, to sum up, I think the problem is still there. The wallet is stuck loading in the same place, the output of the 10M sat channel in limbo is still locked in the address. If I run `lncli pendingchannels` the output is:

`[lncli] rpc error: code = Unknown desc = unable to find arbitrator`

Appreciate the help and the patience with this one, it's a tricky one.

**Reply by guggero** [reply]
No, if you don't specify `--publish`, it won't be published. But you can copy the transaction hex string and publish it manually on a block explorer.

Hmm, so yeah, if Thunderhub said it was a local force close, then `sweeptimelockmanual` is the correct one.
Would you mind sending me the full output of `chantools dumpbackup` (just for this specific channel) and the exact parameters you use for `sweeptimelockmanual` currently? You can also DM me the info on Slack or Keybase (same username) if you want.
